### ISSUE TITLE
Generic Wait logic for Azure Jobs and Provisioning... and beyond!

### ISSUE THREAD

**Issue by MiYanni** [Azure/azure-powershell#5183]
### Cmdlet(s)
[Wait-AzureRmBackupJob](https://docs.microsoft.com/en-us/powershell/module/azurerm.backup/wait-azurermbackupjob)
[Wait-AzureRmDataLakeAnalyticsJob](https://docs.microsoft.com/en-us/powershell/module/azurerm.datalakeanalytics/Wait-AzureRmDataLakeAnalyticsJob)
[Wait-AzureRmHDInsightJob](https://docs.microsoft.com/en-us/powershell/module/azurerm.hdinsight/wait-azurermhdinsightjob)
[Wait-AzureRmRecoveryServicesBackupJob](https://docs.microsoft.com/en-us/powershell/module/azurerm.recoveryservices.backup/wait-azurermrecoveryservicesbackupjob)
### Description
Azure has many services that have the concept of Jobs. Currently, there is no coverage of all of these Jobs in Azure. I'm currently using Automation and there is the concept of a RunBookJob. But, this doesn't have a way to Wait on it (via cmdlet). I'm sure there are other jobs like that where there is no Wait cmdlet for it. The proposal is:
- [ ] Make a generic Azure Job interface
- [ ] Make a Wait cmdlet that can wait on any Azure Job

Additionally, there are processes in which you need to wait for a resource to be provisioned. As it seems, most recommendations online create a polling loop, checking on the ProvisioningState. Then, it simply proceeds when it hits a certain state (like Succeeded). The proposal is:
- [ ] Make a Wait cmdlet that works for resource provisioning, **OR**

You could even go one step further. Here is someone waiting on a VM to restart:
https://gallery.technet.microsoft.com/scriptcenter/Create-a-Wait-State-Loop-f020c2b3
You could take the provision state cmdlet proposal above and make it a generic Wait cmdlet that takes the property you want to wait on as a parameter, and the success state of that property as a parameter. Then, it would even work with his PowerState waiting! 
- [ ] Make a Wait cmdlet that waits on any 'state' property provided and allows you to set the 'break' value of that property

**Comment by alexandair** [comment]
I like your ideas for a generic Wait logic.
Btw, Start-Azure​RmAutomation​Runbook has the Wait parameter.

**Comment by MiYanni** [comment]
@alexandair Thanks for letting me know that Start has -Wait. It is quite convoluted, though. If you don't -Wait, the cmdlet returns a Job object. However, you can't wait on this job object. Meaning, if I want to start a bunch of Runbooks (not using -Wait) and wait for all of them to finish, I can't. That's standard Async design. Tell a bunch of stuff to run and wait for it all to finish. With the current design, they'd have to allow the Start command to accept multiple Runbooks (by name, for example) so that you could start multiple and -Wait on all of them. The only way to achieve this currently would be to run the Start command with -Wait asynchronously using PS Job logic, and wait on all those Start calls to finish.

**Comment by markcowl** [comment]
@MiYanni Do these cmdlets return a ```Job``` object that can be used with ```Wait-Job```  (and the other job-related cmdlets)?  This is the standard for new cmdlets but these cmdlets pre-date the standard.

If not, let's change the issues to make these cmdlets fit the standard (default = wait, -AsJob = return Job)

**Comment by MiYanni** [comment]
@markcowl 
The Job object that can be used by `Wait-Job` is either `System.Management.Automation.Job` or `System.Management.Automation.Job2` from my understanding. After looking at the code:
_Note: [ ] names are the root base class_
- Wait-AzureRmBackupJob: **No** [AzureBackupVaultContextObject]
- Wait-AzureRmDataLakeAnalyticsJob: **No** [JobInformationBasic]
- Wait-AzureRmHDInsightJob: **No** [AzureHDInsightJob]
- Wait-AzureRmRecoveryServicesBackupJob: **No** [ObjectBase]
- Start-AzureRmAutomationRunbook: **No** [Microsoft.Azure.Commands.Automation.Model.Job]

There are probably more 'Job'-type commands, or ones that have a Wait mechanism. We'd have to find all of them and use the Composition pattern so that we derive of Job2 but still have the information required for everything else to use them (base class as a private member).

To do the, "let's change the issues to make these cmdlets fit the standard (default = wait, -AsJob = return Job)" would require a lot of breaking changes and removing/modifying many cmdlets.

**Comment by twitchax** [comment]
@markcowl and I will do some interviews on this.

**Comment by twitchax** [comment]
At this point, still poised to do some more research.  Should have results mid July.

**Comment by MiYanni** [comment]
Recently found out that `DataMigrationService` has similar polling logic. Their cmdlet ends in `Task`, so that might be another good search term. Info from this review: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/155

Example:
```powershell
$migTask = New-AzureRmDmsTask -TaskType MigrateSqlServerSqlDbSync `
    -ResourceGroupName $rg.ResourceGroupName `
    -ServiceName $service.Name `
    -ProjectName $project.Name `
    -TaskName $taskName `
    -SourceConnection $sourceConnectionInfo `
    -SourceCred $sourceCred `
    -TargetConnection $targetConnectionInfo `
    -TargetCred $targetCred `
    -SelectedDatabase  $selectedDbs

$task = Get-AzureRmDataMigrationTask -ResourceGroupName $rg.ResourceGroupName -ServiceName $service.Name -ProjectName $project.Name -TaskName $taskName -Expand

while(($task.ProjectTask.Properties.State -eq "Running") -or ($task.ProjectTask.Properties.State -eq "Queued")) {
	Foreach($output in $syncTask.ProjectTask.Properties.Output) {
	        if ($output.Id -clike "db|*") {
			Write-Host ($output | Format-List | Out-String)

			if ($output.MigrationState -eq "READY_TO_COMPLETE") {
				$command = New-AzureRmDmsCommand -CommandType Complete `
					-ResourceGroupName $rg.ResourceGroupName `
					-ServiceName $service.Name `
					-ProjectName $project.Name `
					-TaskName $taskName `
					-DatabaseName $output.DatabaseName`
			}
                }
        }

	SleepTask 15
	$task = Get-AzureRmDataMigrationTask -ResourceGroupName $rg.ResourceGroupName -ServiceName $service.Name -ProjectName $project.Name -TaskName $taskName -Expand 
}
```

**Comment by teresaritorto** [comment]
This issue has been marked as stale and is being closed due to inactivity. If you are still experiencing this problem, please create a new issue and reference the latest version of the Azure PowerShell tool. This helps us ensure that we address current scenarios and provide accurate support. Thank you for your understanding!
